types/mempool/mempool.go

```diff
@@ -0,0 +1,254 @@
+)
+
+// Tx we define an app-side mempool transaction interface that is as
+// minimal as possible, only requiring applications to define the size of the
+// transaction to be used when reaping and getting the transaction itself.
+// Interface type casting can be used in the actual app-side mempool implementation.
+type Tx interface {
+	types.Tx
+
+	// Size returns the size of the transaction in bytes.
+	Size() int64
+}
+
+type Mempool interface {
+	// Insert attempts to insert a Tx into the app-side mempool returning
+	// an error upon failure.
+	Insert(types.Context, Tx) error
+
+	// Select returns the next set of available transactions from the app-side
+	// mempool, up to maxBytes or until the mempool is empty. The application can
+	// decide to return transactions from its own mempool, from the incoming
+	// txs, or some combination of both.
+	Select(txs [][]byte, maxBytes int64) ([]Tx, error)
+
+	// CountTx returns the number of transactions currently in the mempool.
+	CountTx() int
+
+	// Remove attempts to remove a transaction from the mempool, returning an error
+	// upon failure.
+	Remove(Tx) error
+}
+
+type ErrTxNotFound struct {
+	error
+}
+
+type Factory func() Mempool
+
+var _ Mempool = (*defaultMempool)(nil)
+
+// defaultMempool is the SDK's default mempool implementation which stores txs in a partially ordered set
+// by 2 dimensions; priority, and sender-nonce.  Internally it uses one priority ordered skip list and one skip list
+// per sender ordered by nonce (sender-nonce).  When there are multiple txs from the same sender, they are not
+// always comparable by priority to other sender txs and must be partially ordered by both sender-nonce and priority.
+type defaultMempool struct {
+	priorities *huandu.SkipList
+	senders    map[string]*huandu.SkipList
+	scores     map[txKey]int64
+	iterations int
+}
+
+type txKey struct {
+	nonce    uint64
+	priority int64
+	sender   string
+}
+
+// txKeyLess is a comparator for txKeys that first compares priority, then sender, then nonce, uniquely identifying
+// a transaction.
+func txKeyLess(a, b interface{}) int {
+	keyA := a.(txKey)
+	keyB := b.(txKey)
+	res := huandu.Int64.Compare(keyA.priority, keyB.priority)
+	if res != 0 {
+		return res
+	}
+
+	res = huandu.String.Compare(keyA.sender, keyB.sender)
+	if res != 0 {
+		return res
+	}
+
+	return huandu.Uint64.Compare(keyA.nonce, keyB.nonce)
+}
+
+// NewDefaultMempool returns the SDK's default mempool implementation which returns txs in a partial order
+// by 2 dimensions; priority, and sender-nonce.
+func NewDefaultMempool() Mempool {
+	return &defaultMempool{
+		priorities: huandu.New(huandu.LessThanFunc(txKeyLess)),
+		senders:    make(map[string]*huandu.SkipList),
+		scores:     make(map[txKey]int64),
+	}
+}
+
+// Insert attempts to insert a Tx into the app-side mempool in O(log n) time, returning an error if unsuccessful.
+// Sender and nonce are derived from the transaction's first signature.
```

Comment:
So this is an unfortunate case. First, you have to recognize that most txs are single message, i.e. one signer, even multi-sig msgs are one signer. 

Due to relayers, we now also have many multi-msg txs. But even those are still single signer, i.e. the relayer account.

In other words, 99% of txs are single singer, regardless of the number of messages.

Now, it can totally be possible that a multi-msg tx is constructed with _different_  signers, but this hardly, if ever, happens in practice. There's just no practical use case or reason for it.

That being said, it still is a case that we unfortunately must consider. So I'm not 100% sure how this implementation handles it. But one thing I can think of immediately is that whenever multiple signers exist (> 1), then we just default to 0 priority or maybe put them in a special "bucket/list" and only pick from that list when the primary indexes are empty...idk. We could also leave this undefined for now.

So TL;DR single signer txs are pretty much 99% of all txs, if not all.

Reply:
^^^ This sums it up well.  Since multiple signers/nonces are a rare (I can't find a historical tx example, if anyone reading can please share it) and I have no direct knowledge of that use case we chose to always use the first signature to derive nonce and sender.   Even in [this case](https://www.mintscan.io/cosmos/txs/74996D72120820974FE125B15537D0E6BA11980D16F1934BB1849419949791BF) of multiple PubKeys because there is a single `signer_info` it all rolls up under one signer account and sequence number.

In this implementation there is no special casing for multiple `signer_info`s, the first one will be used and then the tx prioitized by there.  Is there a reason to treat them different, i.e. deprioritized like your first suggestion @alexanderbez?